Sources/GRPC/AsyncAwaitSupport/GRPCAsyncServerHandler.swift

```diff
@@ -0,0 +1,689 @@
+      self.interceptors.send(.metadata([:]), promise: nil)
+
+      // Spin up a task to call the async user handler.
+      self.userHandlerTask = userHandlerPromise.completeWithTask {
+        try await withTaskCancellationHandler {
+          do {
+            // Call the user function.
+            try await self.userHandler(requestStream, responseStreamWriter, context)
+          } catch {
+            responseStreamWriter._asyncWriter.cancelAsynchronously()
+            // Throwing GRPCStatus.ok is considered to be invalid.
+            if (error as? GRPCStatus)?.isOk ?? false {
+              throw GRPCStatus(
+                code: .unknown,
+                message: "Handler threw GRPCStatus error with code .ok"
+              )
+            }
+            throw error
+          }
+        } onCancel: {
+          /// The task being cancelled from outside is the signal to this task that an error has
+          /// occured and we should abort the user handler.
+          ///
+          /// Adopters are encouraged to cooperatively check for cancellation in their handlers but
+          /// we cannot rely on this.
+          ///
+          /// We additionally signal the handler that an error has occured by terminating the source
+          /// backing the request stream that the user handler is consuming.
+          ///
+          /// - NOTE: This handler has different semantics from the extant non-async-await handlers
+          /// where the `statusPromise` was explicitly failed with `GRPCStatus.unavailable` from
+          /// _outside_ the user handler. Here we terminate the request stream with a
+          /// `CancellationError` which manifests _inside_ the user handler when it tries to access
+          /// the next request in the stream. We have no control over the implementation of the user
+          /// handler. It may choose to handle this error or not. In the event that the handler
+          /// either rethrows or does not handle the error, this will be converted to a
+          /// `GRPCStatus.unknown` by `handleError(_:)`. Yielding a `CancellationError` _inside_
+          /// the user handler feels like the clearest semantics of what we want--"the RPC has an
+          /// error, cancel whatever you're doing." If we want to preserve the API of the
+          /// non-async-await handlers in this error flow we could add conformance to
+          /// `GRPCStatusTransformable` to `CancellationError`, but we still cannot control _how_
+          /// the user handler will handle the `CancellationError` which could even be swallowed.
+          ///
+          /// - NOTE: Currently we _have_ added `GRPCStatusTransformable` conformance to
+          /// `CancellationError` to convert it into `GRPCStatus.unavailable` and expect to
+          /// document that user handlers should always rethrow `CacellationError` if handled, after
+          /// optional cleanup.
+          requestStreamSource.finish(throwing: CancellationError())
```

Comment:
@glbrntt do you think we should mark `finish(throwing:)` as `@discardableResult`. In this instance I don't think there's anything we can meaningfully do with the `YieldResult`?

Reply:
Done.